File: CPlusPlus/0856_Score_A_Parentheses.cpp
class Solution {
public:
    int scoreOfParentheses(string S) {
        int N = S.size();
        stack<int> s;
        for (int i = 0; i < N; i++) {
            if (S[i] == '(') {
                s.push(-1);
            } else {
                if (!s.empty() && s.top() == -1) {
                    s.pop();
                    s.push(1);
                } else {
                    int local = 0;
                    while (!s.empty() && s.top() != -1) {
                        local += s.top();
                        s.pop();
                    }
                    s.pop();
                    s.push(2 * local);
                }
            }
        }
        int result = 0;
        while (!s.empty()) {
            result += s.top();
            s.pop();
        }
        return result;
    }
};

File: Python/0856_Score_A_Parentheses.py
class Solution:
    def scoreOfParentheses(self, S: str) -> int:
        stack = []
    
        for char in S:
            if char == '(':
                stack.append(char)
            else:
                if stack and stack[-1] == '(':
                    stack.pop()
                    stack.append(1)
                else:
                    local = 0
                    while stack and stack[-1] != '(':
                        local += stack.pop()
                    stack.pop()
                    stack.append(2 * local)
        return sum(stack)
